test/framework/src/main/java/org/elasticsearch/test/InternalTestCluster.java

```diff
@@ -827,6 +832,24 @@ void restart(RestartCallback callback) throws Exception {
                     IOUtils.rm(nodeEnv.nodeDataPaths());
                 }
             }
+            startNewNode(newSettings);
+        }
+
+        void stopNodeWithoutClearingData() throws IOException {
```

Comment:
this is indeed not needed. You can just stop and start and node (it will reused the first available data folder, which is the one the previous node just used and freed). More cleanly, you can use restartRandomDataNode() which gets a callback to executed while the node is offline. In that callback you can delete the index.